docs/specification/new-spec/light-client.md

```diff
@@ -0,0 +1,111 @@
+
+For the purpose of this light client specification, we assume that the Tendermint Full Node exposes the following functions over
+Tendermint RPC:
+
+```golang
+Header(height int64) (SignedHeader, error) // returns signed header for the given height
+Validators(height int64) (ResultValidators, error) // returns validator set for the given height
+LastHeader(valSetNumber int64) (SignedHeader, error)   // returns last header signed by the validator set with the given validator set number
+
+type SignedHeader struct {
+    Header Header 
+    Commit Commit
+    ValSetNumber int64 
+}
+
+type ResultValidators struct {
+    BlockHeight int64              
+    Validators  []Validator 
+    ValSetTime int64               // time the current validator set is initialised, i.e, time of the last validator change before header BlockHeight 
+}
+```
+
+We assume that Tendermint keeps track of the validator set changes and that each time a validator set is changed it is 
+being assigned the next sequence number. We can this number the validator set sequence number. Tendermint also remembers 
+the Time from the header when the next validator set is initialised (starts to be in power), and we refer to this time
+as validator set init time.
+Furthermore, we assume that each validator set change is signed (committed) by the current validator set. More precisely,
+given a block `H` that contains transactions that are modifying the current validator set, the Merkle root hash of the next 
+validator set (modified based on transactions from block H) will be in block `H+1` (and signed by the current validator 
+set), and then starting from the block `H+2`, it will be signed by the next validator set.    
+
+Note that the real Tendermint RPC api is slightly different (for example, response messages contains more data and function 
+names are slightly different); we shortened (and modified) it for the purpose of this document to make the spec more 
+clear and simple. Furthermore, note that in case of the third function, the returned header has `ValSetNumber` equals to 
+`valSetNumber+1`.    
+
+
+Locally, light client manages the following state:
+
+```golang
+valSet []Validator     // current validator set (last known and verified validator set) 
+valSetNumber int64     // sequence number of the current validator set 
+valSetHash []byte      // hash of the current validator set
+valSetTime int64       // time when the current validator set is initialised  
+```
+
+The light client is initialised with the trusted validator set, for example based on the known validator set hash,
+validator set sequence number and the validator set init time.
+The core of the light client logic is captured by the Certify function that is used to 1) verify if the given header is valid,
+and 2) update the validator set (when the given header is valid and it is more recent than the seen headers). 
+
+```golang
+Certify(signedHeader SignedHeader):
```

Comment:
I know this is just a description of the current working of the light client, but `Certify` seems to conflate too many things, validation and mutation namely, while the name does not communicate that.

Reply:
This can be seen more as a description of core mechanisms of the light client. Current implementation is quite different. I agree that this is not the best name and that updating validator set is a side effect. My idea was just to have Certify as an entry point where we show how everything works together. The same underlying mechanisms can be used in the context of IBC, pooled security architecture and the light client. Do you have a suggestion how we organise the content differently?

Reply:
Something that would imply mutation could be a good start e.g.: `Update`